File: java/Tree.java
class Node{
		int data;
		Node left;
		Node right;
		
		public static Node newNode(int no){
			Node n = new Node();
			n.data = no;
			n.left = null;
			n.right = null;
			return n;
		}
	}

public class Tree {
	public Node addNode(int data, Node head){
		Node tempHead = head;
		Node n = Node.newNode(data);
		
		Node prev = null;
		if (head == null){
			head = n;
			return head;
		}
		while(head!=null){
			prev = head;
			if(head.data > n.data){
				head = head.left;
			}else{
				head = head.right;
			}
		}
		if(n.data < prev.data){
			prev.left = n;
		}else{
			prev.right = n;
		}
		return tempHead;
	}
	
	public void inorder(Node root){
		if(root!=null){
			inorder(root.left);
			System.out.println(root.data);
			inorder(root.right);
		}
	}
	
	public static void main(String[] args){
		Tree t = new Tree();
		Node root = null;
		root = t.addNode(5, root);
		root = t.addNode(3, root);
		root = t.addNode(4, root);
		root = t.addNode(1, root);
		root = t.addNode(7, root);
		root = t.addNode(9, root);
		t.inorder(root);
		
//		System.out.println(root.data);
	}	
}

File: 2018/STL/tree.cpp
#include <iostream>
using namespace std;

struct Node {
    int data;
    struct Node *left;
    struct Node *right;
};


struct Node *getNode(int n){
    struct Node *p = (struct Node*)malloc(sizeof(struct Node));
    p->data = n;
    p->left = NULL;
    p->right = NULL;
    return p;
}

void inorder(struct Node *root){
    if(root){
        inorder(root->left);
        cout<<root->data<<endl;
        inorder(root->right);
    }
}

void preorder(struct Node *root){
    if(root){
        cout<<root->data<<endl;
        preorder(root->left);
        preorder(root->right);
    }
}

void postorder(struct Node *root){
    if(root){
        preorder(root->left);
        preorder(root->right);
        cout<<root->data<<endl;
    }
}


int main(){
    struct Node *root = getNode(100);
    root->left = getNode(1);
    root->right = getNode(200);
    root->left->left = getNode(12);
    root->right->right = getNode(22);
    inorder(root);

    return 0;
}